src/main/java/io/cdap/plugin/http/common/pagination/BaseHttpPaginationIterator.java

```diff
@@ -84,13 +84,20 @@ protected boolean visitPageAndCheckStatusCode() throws IOException {
     if (response != null) { // close previous response
       response.close();
     }
+    if (this.shouldRetry()) {
+      getHttpClient().close(); // In case of retry attempt, close the previous HttpClient instance
+    }
 
     response = new HttpResponse(getHttpClient().executeHTTP(nextPageUrl));
     currentPageUrl = nextPageUrl;
     httpStatusCode = response.getStatusCode();
-    RetryableErrorHandling errorHandlingStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode);
 
-    return !errorHandlingStrategy.shouldRetry();
+    return !shouldRetry();
+  }
+
+  private boolean shouldRetry() {
+    RetryableErrorHandling errorHandlingStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode);
```

Comment:
As a general coding practice, may be a good idea to check the object being null, to avoid any null handling exception.

Reply:
The httpErrorHandler is created in the constructor and not set to null any where in the code. Hence it should not be null at this point.